src/Analyzers/Core/Analyzers/NamespaceMatchFolder/AbstractNamespaceMatchFolderDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,196 @@
+
+            // Project directory is a must to correctly get the relative path and construct a namespace
+            if (!context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(ProjectDirOption, out var projectDir)
+                || string.IsNullOrEmpty(projectDir))
+            {
+                return;
+            }
+
+            if (context.Node is TNamespaceSyntax namespaceDecl)
+            {
+                var currentNamespace = GetNamespaceName(context.SemanticModel, namespaceDecl);
+
+                if (IsFileAndNamespaceMismatch(namespaceDecl, rootNamespace, projectDir, currentNamespace, out var targetNamespace) &&
+                    IsFixSupported(context.SemanticModel, namespaceDecl, context.CancellationToken))
+                {
+                    var nameSyntax = GetNameSyntax(namespaceDecl);
+
+                    context.ReportDiagnostic(Diagnostic.Create(
+                        Descriptor,
+                        nameSyntax.GetLocation(),
+                        additionalLocations: null,
+                        properties: ImmutableDictionary<string, string?>.Empty.Add(TargetNamespace, targetNamespace),
+                        messageArgs: new[] { currentNamespace, targetNamespace }));
+                }
+            }
+
+            static string GetNamespaceName(SemanticModel semanticModel, SyntaxNode namespaceSyntax)
+            {
+                var symbol = semanticModel.GetDeclaredSymbol(namespaceSyntax);
+                RoslynDebug.AssertNotNull(symbol);
+
+                return symbol.ToDisplayString(s_namespaceDisplayFormat);
+            }
+        }
+
+        private bool IsFixSupported(SemanticModel semanticModel, TNamespaceSyntax namespaceDeclaration, CancellationToken cancellationToken)
+        {
+            var root = namespaceDeclaration.SyntaxTree.GetRoot(cancellationToken);
+
+            // It should not be nested in other namespaces
+            if (namespaceDeclaration.Ancestors().OfType<TNamespaceSyntax>().Any())
+            {
+                return false;
+            }
+
+            // It should not contain a namespace
+            var containsNamespace = namespaceDeclaration
+                 .DescendantNodes(n => n is TNamespaceSyntax)
+                 .OfType<TNamespaceSyntax>().Any();
+            if (containsNamespace)
+            {
+                return false;
+            }
+
+            // The current namespace should be valid
+            var isCurrentNamespaceInvalid = GetNameSyntax(namespaceDeclaration)
+                .GetDiagnostics().Any(d => d.Severity == DiagnosticSeverity.Error);
+            if (isCurrentNamespaceInvalid)
+            {
+                return false;
+            }
+
+            // It should not contain partial classes with more than one instance in the semantic model. The
+            // fixer does not support this scenario.
```

Comment:
seems like an odd limitation in teh fixer.  i'm honestly not sure why this wouldn't work...

Reply:
It's still a limitation. If we need it fixed it should be filed as a separate issue because it will impact the sync namespace refactoring

Reply:
sgtm.